src/metalookup/features/addblock_based.py

```diff
@@ -1,9 +1,127 @@
-from core.extractor import ExtractionMethod, MetadataBase, ProbabilityDeterminationMethod
-
-
-@MetadataBase.with_key()
-class Advertisement(MetadataBase):
-    urls = [
+import asyncio
+import logging
+from concurrent.futures import Executor
+from logging import Logger
+from typing import Optional
+from unittest import mock
+
+import adblockparser
+from aiohttp import ClientSession
+
+from metalookup.app.models import Explanation, StarCase
+from metalookup.core.extractor import Extractor, download_tag_lists
+from metalookup.core.website_manager import WebsiteData
+from metalookup.lib.tools import runtime
+
+_FOUND_LIST_MATCHES = "Found list matches"
+_FOUND_NO_LIST_MATCHES = "Found no list matches"
+
+
+class AdBlockBasedExtractor(Extractor[set[str]]):
+    """
+    Base class for features to be extracted with `adblockparser` package.
+
+    The returned extra data corresponds to a sample of the blocked links.
+    The sample size can be configured with the limit member variable.
+    """
+
+    urls: list[str] = []  # needs to be provided by derived class
+
+    def __init__(self, logger: Optional[Logger] = None):
+        self.logger = logger or logging.getLogger(__name__)
+        self.adblock_parser_options = {
+            "script": True,
+            "image": True,
+            "stylesheet": True,
+            "object": True,
+            "xmlhttprequest": True,
+            "object-subrequest": True,
+            "subdocument": True,
+            "document": True,
+            "elemhide": True,
+            "other": True,
+            "background": True,
+            "xbl": True,
+            "ping": True,
+            "dtd": True,
+            "media": True,
+            "third-party": True,
+            "match-case": True,
+            "collapse": True,
+            "donottrack": True,
+            "websocket": True,
+            "domain": "",
+        }
+        self.rules: AdblockRules = None  # noqa will be initialized in async setup call
+        self.limit: Optional[int] = 2
+        """
+        Stop after finding a given amount of to be blocked links.
```

Comment:
This comment's position confused me directly, since I thought it would belong to something coming afterwards, which is the case for most of your comments.
I get the impression you like comments a lot. In my experience comments do not age well, because most people - including me - do not maintain them well. Therefore I personally prefer to have speaking variable names that require no or little commenting. Such a big block for one number feels far too much.

Reply:
In `__init__` and some other places a python string that follows the variable documents the public API: https://stackoverflow.com/a/9558703/2160256

Yes, one has to maintain comments together with the code, and yes, they tend to sometimes be outdated, and yes a good variable names often make a comment obsolete.
But - especially for API documentation like this which may show up in e.g. autogenerated API docs, or your IDE popup - those are not reasons not to comment the public API.

TBH, one of the biggest pain points I had with the original setup of this codebase is that there were almost no comments anywhere, which left me with no clue why stuff was done how it was done...

Reply:
Interesting, attribute docstrings are a thing then. Ok good to know.
Ups :D so basically I refrained from all comments which removed the need to maintain them but caused issues in understanding :/ That is not good.
Please keep the comment here :)